### PR TITLE
New Resource: `azurerm_data_factory_linked_service_odata`

### PR DESCRIPTION
**Pull request by pomverte**
Add support for Terraform configuration for Azure Data Factory Linked Service of type OData.

New resource `azurerm_data_factory_linked_service_odata`
```
resource "azurerm_data_factory_linked_service_odata" "example" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  data_factory_name   = azurerm_data_factory.example.name
  url                 = "https://services.odata.org/v4/TripPinServiceRW/People"
}
```
Documentation : https://docs.microsoft.com/en-us/azure/data-factory/connector-odata

First try: https://github.com/terraform-providers/terraform-provider-azurerm/pull/7704

### CODE REVIEW

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	dataFactoryName := id.Path["factories"]
+	name := id.Path["linkedservices"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:167]: could we use a generated ID parser here?

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	dataFactoryName := id.Path["factories"]
+	name := id.Path["linkedservices"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:231]: and here

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+func (t LinkedServiceODataResource) Exists(ctx context.Context, clients *clients.Client, state *terraform.InstanceState) (*bool, error) {
+	id, err := azure.ParseAzureResourceID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+	resourceGroup := id.ResourceGroup
+	dataFactoryName := id.Path["factories"]
+	name := id.Path["linkedservices"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource_test.go:97]: and here

File: website/docs/r/data_factory_linked_service_odata.html.markdown [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Data Factory Linked Service OData. Changing this forces a new resource to be created. Must be unique within a data
+  factory. See the [Microsoft documentation](https://docs.microsoft.com/en-us/azure/data-factory/naming-rules) for all restrictions.
```
Comment (katbyte) [inline website/docs/r/data_factory_linked_service_odata.html.markdown:53]:
```suggestion
* `name` - (Required) Specifies the name of the Data Factory Linked Service OData. Changing this forces a new resource to be created. Must be unique within a data factory. See the [Microsoft documentation](https://docs.microsoft.com/en-us/azure/data-factory/naming-rules) for all restrictions.
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	description := d.Get("description").(string)
+
+	odataLinkedService := &datafactory.ODataLinkedService{
+		Description: &description,
+		Type:        datafactory.TypeOData,
+	}
+
+	odataLinkedService.ODataLinkedServiceTypeProperties = authenticationProperties(d)
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:151]: we can simplify this
```suggestion
	odataLinkedService := &datafactory.ODataLinkedService{
		Description: util.String(d.Get("description").(string)),
		Type:        datafactory.TypeOData,
		ODataLinkedServiceTypeProperties: authenticationProperties(d)
	}
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+		passwordSecureString := datafactory.SecureString{
+			Value: &password,
+			Type:  datafactory.TypeSecureString,
+		}
+		return &datafactory.ODataLinkedServiceTypeProperties{
+			AuthenticationType: datafactory.ODataAuthenticationType(datafactory.ODataAuthenticationTypeBasic),
+			URL:                utils.String(url),
+			UserName:           username,
+			Password:           &passwordSecureString,
+		}
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:261]: we can combine and simplify all this too

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @pomverte - In addtion to the comments i've left inline, the tests are failing with:
```
Test ended in panic.

------- Stdout: -------
=== RUN   TestAccDataFactoryLinkedServiceOData_anon_auth
=== PAUSE TestAccDataFactoryLinkedServiceOData_anon_auth
=== CONT  TestAccDataFactoryLinkedServiceOData_anon_auth

------- Stderr: -------
2021/05/03 19:31:10 [DEBUG] not using binary driver name, it's no longer needed
2021/05/03 19:31:10 [DEBUG] not using binary driver name, it's no longer needed
panic: runtime error: index out of range [0] with length 0

goroutine 469 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/datafactory.authenticationProperties(0xc0036e39d0, 0x5e6aa60)
	/opt/teamcity-agent/work/a73be106926a7472/azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:249 +0x485
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/datafactory.resourceArmDataFactoryLinkedServiceODataCreateUpdate(0xc0036e39d0, 0x5656440, 0xc00118a000, 0x0, 0x0)
	/opt/teamcity-agent/work/a73be106926a7472/azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:151 +0x22b
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc001845860, 0xc001d43860, 0xc0016a4940, 0x5656440, 0xc00118a000, 0x5685a01, 0xc00327d7c8, 0xc001a5c1e0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:320 +0x375
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc0030cd200, 0xc000e6fa38, 0xc001d43860, 0xc0016a4940, 0xc001a5a088, 0xc0032d9f90, 0x5688ce0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc0019216f8, 0x65f27f0, 0xc001a52ba0, 0xc0036e3260, 0xc0019216f8, 0xc001a52ba0, 0xc001b00ba0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:895 +0x8a5
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x5b6a4a0, 0xc0019216f8, 0x65f27f0, 0xc001a52ba0, 0xc002400f60, 0x0, 0x65f27f0, 0xc001a52ba0, 0xc001a56700, 0x359)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3305 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00127f6c0, 0x6639398, 0xc003058900, 0xc001250f00, 0xc001ccbc50, 0x9a5e6e0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/google.golang.org/grpc/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00127f6c0, 0x6639398, 0xc003058900, 0xc001250f00, 0x0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/google.golang.org/grpc/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000d5c130, 0xc00127f6c0, 0x6639398, 0xc003058900, 0xc001250f00)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/google.golang.org/grpc/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/a73be106926a7472/vendor/google.golang.org/grpc/server.go:857 +0x1fd
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go (hunk 1 above)
Comment (pomverte) [reply azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:200]: Are you referring to something like `uuid.FromStringOrNil(...)` or `uuid.Parse(...)` ?

Comment (pomverte) [reply azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go]: Took me a while to understand what we were talking about ...

The parser had already been generated [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/datafactory/resourceids.go#L4)

But the static type checking is awesome !

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go (hunk 2 above)
Comment (pomverte) [reply azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go]: done

File: azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource_test.go (hunk 3 above)
Comment (pomverte) [reply azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource_test.go]: done

**Review by katbyte — CHANGES_REQUESTED** [review]
Looks like the tests are failing: 
```
est ended in panic.

------- Stdout: -------
=== RUN   TestAccDataFactoryLinkedServiceOData_basic_update
=== PAUSE TestAccDataFactoryLinkedServiceOData_basic_update
=== CONT  TestAccDataFactoryLinkedServiceOData_basic_update

------- Stderr: -------
2021/05/18 20:43:14 [DEBUG] not using binary driver name, it's no longer needed
2021/05/18 20:43:14 [DEBUG] not using binary driver name, it's no longer needed
panic: runtime error: index out of range [0] with length 0

goroutine 484 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/datafactory.authenticationProperties(0xc001b89b20, 0x5f2f76a)
	/opt/teamcity-agent/work/a73be106926a7472/azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:275 +0x425
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/datafactory.resourceArmDataFactoryLinkedServiceODataCreateUpdate(0xc001b89b20, 0x570c9e0, 0xc0002aa300, 0x0, 0x0)
	/opt/teamcity-agent/work/a73be106926a7472/azurerm/internal/services/datafactory/data_factory_linked_service_odata_resource.go:148 +0x27d
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc002942000, 0xc0018f0af0, 0xc000b4cd80, 0x570c9e0, 0xc0002aa300, 0x573c401, 0xc001f12c28, 0xc00242e840)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:320 +0x375
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc003656780, 0xc0013bfa38, 0xc0018f0af0, 0xc000b4cd80, 0xc002c9ebc8, 0xc0021fa3a0, 0x573f640)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc001d268d8, 0x66c84f0, 0xc00216cea0, 0xc001b89500, 0xc001d268d8, 0xc00216cea0, 0xc001909ba0)
	/opt/teamcity-agent/work/a73be106926a7472/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:895 +0x8a5
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x5c28400, 0xc001d268d8, 0x66c84f0, 0xc0
```

**Review by mbfrahry — APPROVED** [review]
LGTM! 

I just fixed the crashes and moved some of the logic around so it'll be easier to expand on if we need to enable some more features. Thanks for opening this PR!

**Pull request merged**